clustering/clustering_solution.py

```diff
@@ -0,0 +1,310 @@
+
+# %%
+df[21] = df[21].apply(lambda x: horsepower_avg if x == '?' else x)
+
+# %%
+df[21] = df[21].astype(float)
+
+# %%
+X = df[[25, 21]]
+
+# %%
+plt.scatter(df[25], df[21])
+
+# %%
+inertia = []
+s_score = [] # Silhouette: Maximize this as close to 1
+c_score = [] # Calinski-Harabasz: Maximize whichever's highest
+
+for k in range (2, 11): # For k values between 2 and 10
+    km = KMeans(n_clusters = k, n_init = 20)
+    km.fit(X)
+    inertia.append(km.inertia_)
+    s_score.append(silhouette_score(X, km.labels_))
+    c_score.append(calinski_harabasz_score(X, km.labels_))
+
+# %%
+inertia
+
+# %%
+s_score
+
+# %%
+c_score
+
+# %%
+plt.plot(range(2, 11), inertia)
+
+# %%
+plt.plot(range(2, 11), s_score)
+
+# %%
+plt.plot(range(2, 11), c_score)
+
+# %%
+# Let's try k = 3
+
+model = KMeans(n_clusters = 3)
+model.fit_transform(X)
+
+# %%
+plt.scatter(X[25], X[21], c = model.labels_)
+plt.scatter(model.cluster_centers_[:,0], model.cluster_centers_[:,1], marker = 'x', c = 'r')
+
+# %%
+mms = MinMaxScaler()
+X_scaled = pd.DataFrame(mms.fit_transform(X), columns = X.columns)
+X_scaled
+
+# %%
+inertia_scaled = []
+s_score_scaled = [] # Silhouette: Maximize this as close to 1
+c_score_scaled = [] # Calinski-Harabasz: Maximize whichever's highest
+
+for k in range (2, 11): # For k values between 2 and 10
+    km_scaled = KMeans(n_clusters = k, n_init = 20)
+    km_scaled.fit(X_scaled)
+    inertia_scaled.append(km_scaled.inertia_)
+    s_score_scaled.append(silhouette_score(X_scaled, km_scaled.labels_))
+    c_score_scaled.append(calinski_harabasz_score(X_scaled, km_scaled.labels_))
+
+# %%
+model_scaled = KMeans(n_clusters = 3)
+model_scaled.fit_transform(X_scaled)
+
+# %%
+plt.scatter(X_scaled[25], X_scaled[21], c = model_scaled.labels_)
+plt.scatter(model_scaled.cluster_centers_[:,0], model_scaled.cluster_centers_[:,1], marker = 'x', c = 'r')
+
+# %% [markdown]
+# # 2. Using 'price' and 'horsepower' columns from import-85.data, generate a DBSCAN model.
+
+# %% [markdown]
+# #### Plot your results to visualize outliers.
+
+# %%
+dbs = DBSCAN(eps = 0.1057825, min_samples = 7)
+dbs.fit(X_scaled)
+
+# %%
+plt.scatter(X_scaled[25], X_scaled[21], c = dbs.labels_)
+
+# %%
+dbs.labels_
+
+# %%
+labs = dbs.labels_
+
+# %%
+labs[dbs.core_sample_indices_] = 1
+
+# %%
+labs
+
+# %%
+plt.scatter(X_scaled[25], X_scaled[21], c = labs)
+
+# %% [markdown]
+# # 3. Using mlb_batting_cleaned.csv, write a function that takes a player's name and shows the 2 closest players using the nearest neighbors algorithm.
+
+# %%
+mlb = pd.read_csv(r'C:\Users\trucn\Documents\repositories\ds-mod-2-0924-code-solutions\clustering\clustering\mlb_batting_cleaned.csv')
+
+# %%
+# Try utilizing np.where somewhere in thi
+
+mlb
+
+# %%
+mlb2 = mlb.copy()
+
+# %%
+# OHE Tm column
+ohe = ce.OneHotEncoder(use_cat_names = True)
+ohe.fit(mlb2['Tm'])
+tm_ohe = ohe.transform(mlb2['Tm'])
+
+# %%
+# OHE Lg column
+ohe.fit(mlb2['Lg'])

```

Comment:
I think it's usually best practice to either fit/transform multiple columns at once or use two separate OHEs.